src/linklabel.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright 2018 Google LLC
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//! Link label parsing and matching.
+
+use std::borrow::Cow;
+
+use unicase::UniCase;
+
+use crate::scanners::{scan_ch, is_ascii_whitespace, scan_while};
+
+pub type LinkLabel<'a> = UniCase<Cow<'a, str>>;
+
+pub(crate) fn scan_link_label<'a>(text: &'a str) -> Option<(usize, LinkLabel<'a>)> {
+    if scan_ch(text, b'[') == 0 {
+        return None;
+    }
+    let mut i = 1;
+    let mut only_white_space = true;
+    let mut still_borrowed = true;
+    let bytes = text.as_bytes();
+    // no worries, doesnt allocate until we push things onto it
+    let mut label = String::new();
+
+    // TODO: loop over chars instead of bytes to be more utf-8 compatible?
```

Comment:
I think so, see below. The way commonmark is designed, you almost never need to look at chars instead of bytes, pretty much everything non-ASCII is a pass-through. But this is probably the exception.